internal/backoff/backoff.go

```diff
@@ -71,3 +72,32 @@ func (bc Exponential) Backoff(retries int) time.Duration {
 	}
 	return time.Duration(backoff)
 }
+
+// RunF provides a convenient way to run a function f with a caller provided
+// backoff. It runs f repeatedly, backing off when f returns false and resetting
+// the backoff state when f returns true, until the context expires or f returns
+// a non-nil error.
+func RunF(ctx context.Context, f func() (bool, error), backoff func(int) time.Duration) {
```

Comment:
Another maybe bad idea: what if `backoff`, which is currently a `func(int) time.Duration`, was actually a named type that had a `Run[F]` method that did this?

Reply:
We do have a single method interface here which wraps this method: https://github.com/grpc/grpc-go/blob/4ced601604bbe2eede1e73b1b4cecb37c26c2bae/internal/backoff/backoff.go#L34

I initially considered passing this interface as the third parameter to the `RunF` function, but figured that all the callsites currently simply have a field of type `func (int) time.Duration` and not a `backoff.Strategy`. So, that would have required unnecessary wrapping at the callsites. Do you think that might an idea worth chasing?

Reply:
Hmm, no I think this is fine, and if we want to rework things in the future, we always can make more changes.